src/components/Channel/Channel.js

```diff
@@ -1,266 +1,457 @@
-    if (!this.props.channel || !this.props.channel.cid) {
-      return this.props.EmptyPlaceholder;
+    if (originalTitle.current) {
+      document.title = originalTitle.current;
     }
-    // We use a wrapper to make sure the key variable is set.
-    // this ensures that if you switch channel the component is recreated
-    return <ChannelInner {...this.props} key={this.props.channel.cid} />;
-  }
-}
-
-class ChannelInner extends PureComponent {
-  constructor(props) {
-    super(props);
-    this.state = {
-      error: false,
-      // Loading the initial content of the channel
-      loading: true,
-      // Loading more messages
-      loadingMore: false,
-      hasMore: true,
-      messages: Immutable([]),
-      online: true,
-      typing: Immutable({}),
-      watchers: Immutable({}),
-      members: Immutable({}),
-      read: Immutable({}),
-      eventHistory: {},
-      thread: false,
-      threadMessages: [],
-      threadLoadingMore: false,
-      threadHasMore: true,
+  }, [channel]);
+
+  const markReadThrottled = useCallback(
+    throttle(markRead, 500, { leading: true, trailing: true }),
+    [markRead],
+  );
+
+  const handleEvent = useCallback(
+    (e) => {
+      dispatch({ type: 'updateThreadOnEvent', message: e.message, channel });
+
+      if (e.type === 'message.new') {
+        let mainChannelUpdated = true;
+        if (e.message.parent_id && !e.message.show_in_channel) {
+          mainChannelUpdated = false;
+        }
+
+        if (
+          mainChannelUpdated &&
+          e.message.user.id !== chatContext.client.userID
+        ) {
+          if (!document.hidden) {
+            markReadThrottled();
+          } else {
+            const unread = channel.countUnread(lastRead.current);
+            document.title = `(${unread}) ${originalTitle.current}`;
+          }
+        }
+      }
+
+      throttledCopyStateFromChannel();
+    },
+    [
+      channel,
+      throttledCopyStateFromChannel,
+      chatContext.client.userID,
+      markReadThrottled,
+    ],
+  );
+
+  // useLayoutEffect here to prevent spinner. Use Suspense when it is available in stable release
+  useLayoutEffect(() => {
+    let errored = false;
+    let done = false;
+    const onVisibilityChange = () => {
+      if (!document.hidden) {
+        markRead();
+      }
     };
 
-    // hard limit to prevent you from scrolling faster than 1 page per 2 seconds
-    this._loadMoreFinishedDebounced = debounce(this.loadMoreFinished, 2000, {
-      leading: true,
-      trailing: true,
-    });
+    (async () => {
+      if (!channel.initialized) {
+        try {
+          await channel.watch();
+        } catch (e) {
+          dispatch({ type: 'setError', error: e });
+          errored = true;
+        }
+      }
+      done = true;
+      originalTitle.current = document.title;
+      if (!errored) {
+        dispatch({ type: 'initStateFromChannel', channel });
+        if (channel.countUnread() > 0) channel.markRead();
```

Comment:
Why do we call `markRead()` above and `channel.markRead()` directly here?

Reply:
Good question, I'm not quite sure either. This was how the old component handled things, so I just kept it for the time being.

But since the `markRead` callback does almost the same if called from here with some additional safeguards, it makes sense to just call that instead. I changed that in the latest commit.